File: retrofit/src/main/java/com/xmartlabs/bigbang/retrofit/common/ServiceStringConverter.java
package com.xmartlabs.bigbang.retrofit.common;

import com.annimon.stream.Objects;

import org.threeten.bp.LocalDate;
import org.threeten.bp.LocalDateTime;
import org.threeten.bp.ZoneOffset;

import java.lang.annotation.Annotation;
import java.lang.reflect.Type;

import retrofit2.Converter;
import retrofit2.Retrofit;

/** Converts {@link LocalDate} and {@link LocalDateTime} to Strings */
public class ServiceStringConverter extends Converter.Factory {
  @Override
  public Converter<?, String> stringConverter(Type type, Annotation[] annotations, Retrofit retrofit) {
    if (type instanceof Class) {
      if (Objects.equals(type, LocalDate.class)) {
        return value -> String.valueOf(((LocalDate) value).atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli());
      } else if (Objects.equals(type, LocalDateTime.class)) {
        return value -> String.valueOf(((LocalDateTime) value).toInstant(ZoneOffset.UTC).toEpochMilli());
      }
    }
    return super.stringConverter(type, annotations, retrofit);
  }
}

File: retrofit/src/main/java/com/xmartlabs/bigbang/retrofit/common/ServiceStringConverter.kt
package com.xmartlabs.bigbang.retrofit.common

import org.threeten.bp.LocalDate
import org.threeten.bp.LocalDateTime
import org.threeten.bp.ZoneOffset
import retrofit2.Converter
import retrofit2.Retrofit
import java.lang.reflect.Type

/** Converts [LocalDate] and [LocalDateTime] to Strings  */
class ServiceStringConverter : Converter.Factory() {
  override fun stringConverter(type: Type?, annotations: Array<Annotation>?,
                               retrofit: Retrofit?): Converter<*, String> {
    return when (type) {
      is Class<*> -> when (type) {
        LocalDate::class.java -> Converter { value: LocalDate -> value.atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli().toString() }
        LocalDateTime::class.java -> Converter { value: LocalDateTime -> value.toInstant(ZoneOffset.UTC).toEpochMilli().toString() }
        else -> super.stringConverter(type, annotations, retrofit)
      }
      else -> super.stringConverter(type, annotations, retrofit)
    }
  }
}
